src/Composer/Repository/ComposerRepository.php

```diff
@@ -1017,7 +1017,7 @@ private function loadAsyncPackages(array $packageNames, ?array $acceptableStabil
             $promises[] = $this->startCachedAsyncDownload($name, $realName)
                 ->then(function (array $spec) use (&$packages, &$namesFound, $realName, $constraint, $acceptableStabilities, $stabilityFlags, $alreadyLoaded): void {
                     [$response, $packagesSource] = $spec;
-                    if (null === $response) {
+                    if (null === $response || !isset($response['packages'][$realName])) {
```

Comment:
How about just this? As that would clearly indicate that package data is not provided at all here IMO.

```suggestion
                    if (null === $response || !isset($response['packages'])) {
```

Reply:
Previously if $response['packages'][$realName]  wasn't set $response would be set to null by startCachedAsyncDownload()

However now with the change in startCachedAsyncDownload() if a security advisory is present $response will not be null yet $response['packages'] could  be present while  $response['packages'][$realName] is not present.

The remaining code in that method method relies on the assumption that $response['packages'][$realName] has been validated as being set.

It would be a (badly) broken repository for that to happen, however in my opinion it is better that we validate rather than assume the response contains the key we need.

Reply:
Ok never mind I took a closer look now, thanks for explaining. Looks good to me then, thanks!